File: backend/app/routers/locations.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from sqlalchemy import func

from app.database import get_db
from app.models import Location, Device, User
from app.schemas import LocationCreate, LocationUpdate, LocationOut
from app.auth import get_current_user

router = APIRouter(prefix="/api/locations", tags=["locations"])


def enrich_location(loc: Location, db: Session) -> dict:
    count = db.query(func.count(Device.id)).filter(Device.location_id == loc.id).scalar()
    return {
        "id": loc.id,
        "name": loc.name,
        "description": loc.description,
        "device_count": count or 0,
        "created_at": loc.created_at,
    }


@router.get("", response_model=list[LocationOut])
def list_locations(db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    locations = db.query(Location).order_by(Location.name).all()
    return [enrich_location(loc, db) for loc in locations]


@router.post("", response_model=LocationOut, status_code=201)
def create_location(body: LocationCreate, db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    if db.query(Location).filter(Location.name == body.name).first():
        raise HTTPException(status_code=400, detail="Location name already exists")
    loc = Location(**body.model_dump())
    db.add(loc)
    db.commit()
    db.refresh(loc)
    return enrich_location(loc, db)


@router.put("/{location_id}", response_model=LocationOut)
def update_location(location_id: int, body: LocationUpdate, db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    loc = db.query(Location).filter(Location.id == location_id).first()
    if not loc:
        raise HTTPException(status_code=404, detail="Location not found")
    for field, value in body.model_dump(exclude_unset=True).items():
        setattr(loc, field, value)
    db.commit()
    db.refresh(loc)
    return enrich_location(loc, db)


@router.delete("/{location_id}")
def delete_location(location_id: int, db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    if user.role != "admin":
        raise HTTPException(status_code=403, detail="Admin only")
    loc = db.query(Location).filter(Location.id == location_id).first()
    if not loc:
        raise HTTPException(status_code=404, detail="Location not found")
    db.query(Device).filter(Device.location_id == location_id).update({"location_id": None})
    db.delete(loc)
    db.commit()
    return {"message": "Location deleted"}

File: backend/internal/handler/locations.go
package handler

import (
	"net/http"

	"github.com/yunuskargi/configbox/internal/auth"
	"github.com/yunuskargi/configbox/internal/config"
	"github.com/yunuskargi/configbox/internal/database"
	"github.com/yunuskargi/configbox/internal/models"
	"github.com/yunuskargi/configbox/internal/service"
)

func ListLocations(w http.ResponseWriter, r *http.Request) {
	rows, err := database.DB.Query(`
		SELECT l.id, l.name, l.description, l.created_at, COUNT(d.id) as device_count
		FROM locations l LEFT JOIN devices d ON d.location_id = l.id
		GROUP BY l.id ORDER BY l.name`)
	if err != nil {
		writeError(w, 500, "Database error")
		return
	}
	defer rows.Close()

	var result []models.LocationOut
	for rows.Next() {
		var loc models.LocationOut
		var desc *string
		rows.Scan(&loc.ID, &loc.Name, &desc, &loc.CreatedAt, &loc.DeviceCount)
		loc.Description = desc
		result = append(result, loc)
	}
	if result == nil {
		result = []models.LocationOut{}
	}
	writeJSON(w, 200, result)
}

type locationBody struct {
	Name        string  `json:"name"`
	Description *string `json:"description"`
}

func CreateLocation(w http.ResponseWriter, r *http.Request) {
	user := auth.GetUser(r)
	var body locationBody
	if err := decodeBody(r, &body); err != nil {
		writeError(w, 400, "Invalid request body")
		return
	}

	var exists int
	database.DB.Get(&exists, "SELECT COUNT(*) FROM locations WHERE name = ?", body.Name)
	if exists > 0 {
		writeError(w, 400, "Location name already exists")
		return
	}

	res, _ := database.DB.Exec("INSERT INTO locations (name, description, created_at) VALUES (?, ?, ?)", body.Name, body.Description, config.Now())
	id, _ := res.LastInsertId()

	var loc models.LocationOut
	database.DB.QueryRow(`
		SELECT l.id, l.name, l.description, l.created_at, COUNT(d.id)
		FROM locations l LEFT JOIN devices d ON d.location_id = l.id
		WHERE l.id = ? GROUP BY l.id`, id).Scan(&loc.ID, &loc.Name, &loc.Description, &loc.CreatedAt, &loc.DeviceCount)

	uid := user.ID
	service.LogAction(&uid, user.Username, "create", "location", body.Name, "", clientIP(r))

	writeJSON(w, 201, loc)
}

func UpdateLocation(w http.ResponseWriter, r *http.Request) {
	user := auth.GetUser(r)
	id := paramInt(r, "id")
	var body locationBody
	if err := decodeBody(r, &body); err != nil {
		writeError(w, 400, "Invalid request body")
		return
	}

	var exists int
	database.DB.Get(&exists, "SELECT COUNT(*) FROM locations WHERE id = ?", id)
	if exists == 0 {
		writeError(w, 404, "Location not found")
		return
	}

	if body.Name != "" {
		database.DB.Exec("UPDATE locations SET name = ? WHERE id = ?", body.Name, id)
	}
	if body.Description != nil {
		database.DB.Exec("UPDATE locations SET description = ? WHERE id = ?", *body.Description, id)
	}

	var loc models.LocationOut
	database.DB.QueryRow(`
		SELECT l.id, l.name, l.description, l.created_at, COUNT(d.id)
		FROM locations l LEFT JOIN devices d ON d.location_id = l.id
		WHERE l.id = ? GROUP BY l.id`, id).Scan(&loc.ID, &loc.Name, &loc.Description, &loc.CreatedAt, &loc.DeviceCount)

	uid := user.ID
	service.LogAction(&uid, user.Username, "update", "location", loc.Name, "", clientIP(r))

	writeJSON(w, 200, loc)
}

func DeleteLocation(w http.ResponseWriter, r *http.Request) {
	user := auth.GetUser(r)
	if user.Role != "admin" {
		writeError(w, 403, "Admin only")
		return
	}

	id := paramInt(r, "id")
	var exists int
	database.DB.Get(&exists, "SELECT COUNT(*) FROM locations WHERE id = ?", id)
	if exists == 0 {
		writeError(w, 404, "Location not found")
		return
	}

	var locName string
	database.DB.Get(&locName, "SELECT name FROM locations WHERE id = ?", id)

	database.DB.Exec("UPDATE devices SET location_id = NULL WHERE location_id = ?", id)
	database.DB.Exec("DELETE FROM locations WHERE id = ?", id)

	uid := user.ID
	service.LogAction(&uid, user.Username, "delete", "location", locName, "", clientIP(r))

	writeJSON(w, 200, map[string]string{"message": "Location deleted"})
}
